RichText: Add (foot)notes

ellatrix opened the pull request:
## Description

Fixes #1890.
Blocked by #16428.

![notes](https://user-images.githubusercontent.com/4710635/62622806-8f406e80-b91f-11e9-9574-1be7a7f01bd4.gif)

Front end:

<img width="671" alt="Screenshot 2019-08-07 at 14 31 11" src="https://user-images.githubusercontent.com/4710635/62623000-0a098980-b920-11e9-8536-ef92a81c7196.png">

Notes are not a trivial thing to implement, but they are commonly used, even customary in a lot of fields. I think it’s important that core adds support for them. Adding support for footnotes will also the door for adding comments. They are basically the same thing, just styled differently and only displayed in the editor.

### Positioning

There are three possible ways to position footnotes:

* Endnotes: list at the end of a page.
* In the page margin (possible to do in JS).
* In a popover on hover/click (possible to do in JS and perhaps CSS as well).

“Footnotes” don’t really make sense in the context of a web page, so let's call them "notes".

The first is probably the most common way in print, but the latter seems to become more popular on the web.

It’s worth noting that even though the author may not want to display the notes at the end of the page, the markup should still be there so it can be printed and indexed correctly.

### Markup

There is no standard way to create notes in HTML. Ideally they should have as little markup as possible, so it’s clear, compact and sensible when editing the HTML directly.

There are two parts:

* The note anchor.
* The note list.

The note anchor should obviously be an `a` element and link to the relevant list item. But what should be linked? What should the text be? Any other markup?

A very common thing to do is to “hardcode” the numbering, but that seems too hard to maintain, both manually and programatically. In addition it may depend on the language, script, and convention. The numbering should be added in CSS. Numbering is just a stylistic choice when using endnotes.

Another common thing is to use `sup`, which seems to just be used for convenience. Again, notes might be styled differently depending on your taste, field of work, or language. We should not force a certain style. The markup should be flexible enough to work in a variety of cases.

Notes were originally implemented for print. They where not explicitly linked. Linking is a thing of the web, and we should take it to our advantage as much as possible. E.g. the numbering matters less because of the bidirectional linking. We can also explicitly link the exact text the author wants to reference comment on. If you want to style it differently, fine. You could choose to add a number with CSS `:after`, in any style you like.

This whole idea to only link text starts crumbling down for two reasons though:

* It’s not possible to link anything inside the text you’re commenting on.
* It’s not possible to have multiple references.

So I made it so that you can both link selections of text or insert a reference mark if there is no selection. I do like linked text much more as it’s more explicit than a reference mark, but as said before, it has limitations.

Finally, an `ol` element for the note list makes sense to me to automate the numbering. Inside every list item, a back link should be added that can bring you back to the anchor. Here again, there are different styles: adding it before or after the note. It has never been a thing in print, and there is no standard. Also different characters are used for it. Common ones are ↵ and ↑. In any case, the anchor seems to be in need of an `aria-label`, which should be translatable.

I’m a bit torn on what to do here. We could either save the list in the post content, or dynamically add it to PHP.

I do see the list as derived data, so in that regard it doesn’t make sense to me to store it in the post content. In addition some parts need translation.

But I also see the list as a fallback mechanism if the theme doesn’t have any logic to position them elsewhere. I also don’t really see much of a problem with the list _not_ being dynamically added. Once the post is written and published, it shouldn’t change anyway.

My suggestion is:

```html
<!-- Note for uncollapsed selection -->
Hello <a id="{ id }-anchor" class="note-anchor" href="#{ id }" aria-labelledby="note">world</a>.

<!-- Note for collapsed selection -->
Hello world<a id="{ id }-anchor" class="note-anchor" href="#{ id }" aria-labelledby="note"></a>.

[…]

<!-- Add dynamically or to post content? -->

<footer>
	<h2 id="notes">Notes</h2>
	<ol>
		<li><a id="{ id }" href="#{ id }-anchor" aria-label="?">↑</a> A note about the world.</li>
	</ol>
</footer>
```

<!-- ### Data

The footnote text will be stored in a link attribute. This makes the most sense to me. This way it’s stored right where the link is, which makes it easier to manage changes after moving notes, blocks, etc. The list at the bottom of the page is derived data from the post content.

-->

### First Version

* I decided not to make the note list at the bottom editable for now, but each note _is_ linked to the source where it can be edited. We could add this enhancement in the future if there is demand.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

mtias commented:
> There are three possible ways to position footnotes.

I wonder if this could be a property of the block that shows the footnotes at the bottom.

mtias commented:
This is a great start!

> The numbering should be added in CSS. Numbering is just a stylistic choice when using endnotes.

I tend to agree. Though it does leave the question of "how would this be represented on an email / RSS context?".

> I decided not to make the note list at the bottom editable for now

Bidirectional editing would interesting to have, but agree with keeping it simple at first. Just focusing and opening the relevant footnote in the main content block seems like a good start.

> For now the note can be edited in a plain textarea, but the note can contain HTML. In the future we could put it in a RichText area.

Was there any issue with using RichText there?

ellatrix commented:
> I tend to agree. Though it does leave the question of "how would this be represented on an email / RSS context?".

I've inserted inline CSS styles in the footer, so it should still work. :) Needs to be tested of course.

> Was there any issue with using RichText there?

Yes. `RichText` only works within blocks right now, it relies on a lot of block context. See #15212.

mapk commented:
> The first is probably the most common way in print, but the latter seems to become more popular on the web.

@ellatrix can offer variations for the notes? Can we allow the author to select the preference? Maybe not for version one, but worth keeping in mind for a future iteration. I particularly like displaying them at the bottom of the page to start.

> I’m a bit torn on what to do here. We could either save the list in the post content, or dynamically add it to PHP.

It would be great to see the list appear in the post content at the bottom and be editable. This would be a block that is fixed to the bottom maybe and doesn't allow anything to be added after it? I'm not entirely sure. Some design help is needed here.

> For now the note can be edited in a plain textarea, but the note can contain HTML. In the future we could put it in a RichText area.

I like this. Often times notes should be allowed to link off to the source.

ellatrix commented:
I discussed this PR quite a bit with @mcsf. We didn't really reach any definite conclusion, but here are some thoughts:

> Can we allow the author to select the preference?

I don't think we should. I think it should be a theme thing.

> We could either save the list in the post content, or dynamically add it to PHP.

I think the footnotes should be dynamical generated in PHP, defaulting to accumulating before any "next page" tags, and the end of the content. We could allow filtering, e.g. you could accumulate them before each h2 or something like that. The dynamic generation means that we need to do some HTML matching for the notes, similar to how WordPress core currently replaces static images with dynamic sources etc.

I think I now know enough to be able to finish the PR.

karmatosed commented:
@ellatrix I'd love to give some design feedback once you've iterated a bit, as that sounds like your next step.

ellatrix commented:
@karmatosed It's welcome at any point. :)

karmatosed commented:
This is really great to interact with, huge props for all the work so far. In walking through I do have a few bits of feedback:

- Should note in toolbar be called footnotes? It took me a moment to find it.
- The hover in content cursor seems to not be that of a link. It is in the footer.
- I wonder if the note input could be larger? I don't see why it should be so narrow if the output area isn't.
- I really think there should be an easier way of removing the footnotes, it really took me a little time to work out I had to go into the toolbar and then unclick. Could it be like the link and if you are over an area with a note show in toolbar to 'unlink'?

ellatrix commented:
@karmatosed Thanks for the feedback.

* Since they could be displayed anywhere by the theme, I decided to name it more generally. E.g. the theme could decide to have a pop up for the note on click/hover instead of rendering it at the bottom of the page. Perhaps "note" is slightly too general? But I like it... I'm can't find a good other name that doesn't include "foot". https://en.wikipedia.org/wiki/Note_(typography)
* I'm not sure what you mean here.
* Yes, good point, I'll make it bigger. :) I have some ideas to make it look better.
* I'm not entirely sure what you have in mind. You can just remove the note by pressing backspace? This was obvious in my mind, but it doesn't seem to be to others. I can also add a button to remove them.

karmatosed commented:
> I'm not sure what you mean here.

Sorry for the confusion. When I hover over the footnote link in text it has a different cursor to the one in the footer going to the source. It's hard to get a screenshot of but I can take a video if that doesn't explain.

> I'm not entirely sure what you have in mind. You can just remove the note by pressing backspace? This was obvious in my mind, but it doesn't seem to be to others. I can also add a button to remove them.

Oh! This could totally be me but I didn't even think that would work. I'd be curious to see if that was just my mental model.

ellatrix commented:
I'm going to redo this PR a bit:

* Store footnotes as post meta. This makes more sense to be as it would avoid us storing the data twice. The only thing that will be stored in the post content is an anchor with an ID for the note. Storing as meta also means that this can be reused properly for editorial comments (which we should not store in the post content at all).
* When inserting a note, we create a new ID and either let the user edit the note at the bottom, or move the not closer by so the the user can edit it without scrolling the page. Either way, it's edited in the same component at the bottom (or anywhere else on the page), which updates the meta.

ellatrix commented:
* On the front-end, we match the notes in the content and generate the markup per content page (if there are any created by the "nextpage" blocks) and at the end of the post content. Generating the content is needed so paged posts work properly, and so text and labels can be dynamically translated.

ellatrix commented:
I’ve been working quite a bit on trying to fix footnotes… but I’m reaching some frustrating dead ends.

Since we would like to save footnotes in the `post_content`, some kind of “invisible” block at the end of the post seems to make sense. At first, I thought it would be faster to deduce the footnote *order* from the DOM, because it was fast and easy to do. Later I thought this is a bit messy when you think about history and undo: if footnotes order is calculated at a `componentDidUpdate`, the change is not caught in the same change that updated the order (e.g. moving blocks, inserting a footnote). That’s very annoying. Should footnotes, in general, be recorded in the history? It would be nice…

So better way seems to be to deduce the footnotes *order* from the blocks store… Sounds a bit slower maybe, but it’s just a matter of putting all block attributes in the right order and matching footnote markup in attributes that are strings. That first part is not so easy though… Block attributes themselves are not ordered (imagine two rich text fields in one block) and you don’t know where the nested blocks are rendered relative to those other attributes. Deducing footnotes from the blocks store is a no-go.

Finally, we could also deduce the footnote order from serialised blocks. This seems  _a lot_ slower… and could we ever make it work to have footnotes preserved properly in history (undo)? And I haven’t even begun about pagination of posts. :slightly_smiling_face: What are all your thought? How do you all think footnotes should work? After all these trials and errors with footnotes, I feel like I want to do something else for a moment. :smile:

(I also briefly tried storing footnotes as post meta, as you can see on the branch here, and generate them in PHP on the front-end, but it felt better to have them stored in the `post_content`, so they are always preserved with the rest of the post, and so history would work properly…)

ellatrix commented:
Reworked this a bit. The only major thing left to do now is to add the footnotes on serialisation (save). The footnote order can be determined from the HTML, then added after each page break or at the bottom. Before block parsing they also need to be taken away and added to the store.

LukaszJaro commented:
Awesome!, I was just attempting to implement footnotes for the first time on a new website, I have to take accessibility into account, through my research I discovered this ticket, seems very promising.

Is this feature being pushed into any of the upcoming WP releases soon?

mkaz commented:
Excited for this block! It is possible to make a margin note without JS, by using CSS to float and pull the note right of the content. I have an example in a Sidenote block I created here: https://github.com/mkaz/proximo-sidenote

However, there are complications for mobile and narrow displays, as you mention above, this could be handled by the theme customization. 

There are additional use cases worth considering, beyond adding a text footnote. The sidenote block was inspired by Tufte CSS which allows for a rich side note which may include images, text, links, etc... 

I'm not sure if that scope is too much for what you are aiming for, it could be considered for a v2 or for a plugin to extend this core feature.

aduth commented on docs/designers-developers/developers/data/data-core-editor.md:
```diff
@@ -382,6 +382,18 @@ _Related_
 
 -   getFirstMultiSelectedBlockClientId in core/block-editor store.
 
+<a name="getFootnotes" href="#getFootnotes">#</a> **getFootnotes**
```
Is this something we could use custom sources for (#16402), rather than to build-in awareness of footnotes to the editor module?

ellatrix replied:
I don't have a super deep understanding of custom sources. Would it work if the source is the post content?

mapk commented:
I just checked this out today as well. I'm loving footnotes and can't wait to use it regularly in my personal blog!

1. Can we show the entire footnote in the editor? I was surprised to see it cut off.

![footnote](https://user-images.githubusercontent.com/617986/61835389-550ba300-ae30-11e9-8057-24c62d484f95.gif)

----
2. Does it make sense to be able to add a link to a Footnote? Doesn't it already _link_ to the footnote?

![Screen Shot 2019-07-24 at 4 30 40 PM](https://user-images.githubusercontent.com/617986/61835422-779dbc00-ae30-11e9-8e98-c84975fd72dd.png)

----
3. The footnotes appear a bit too close to the block above and too close to the page edge below. Can we add more padding or margin there? An all-around padding of 40px feels much more comfortable.

ellatrix commented:
@mapk The PR is not really ready for a full UX review. A lot of stuff's been moved around and I still need to polish things. I used input fields for now, I hope that we can use rich text later on.

mcsf commented on packages/block-library/src/footnotes/save.js:
```diff
@@ -0,0 +1,29 @@
+					<a
+						id={ id }
+						href={ `#${ id }-anchor` }
+						aria-label={ __( 'Back to content' ) }
```
This will be tricky until we have better validation for blocks if a user edits the post with a different locale loaded. Until we get there, how realistically can we a) hardcode a string in English (in Esperanto?) instead of passing it through `__`; or b) can we drop the `aria-label` attribute? More interestingly, is there any accessible standard or convention that indicates the notion of "returning" that is language-independent?

ellatrix replied:
Yeah, that's annoying... Emoji? Are they read and translated?

ellatrix replied:
This was also a reason against saving the footnotes in the post content, so it can be generated in PHP.

mcsf commented:
A question concerning UX/IA: Should the Document Outline reveal the Footnotes block? Should the Block Navigation?

Footnotes are a virtual block of sorts, so it's not obvious where the line would be drawn between counting them as a block and not. But I ask because I found myself opening the Outline tray _expecting_ to find them there.

ellatrix committed:
Squash:
197f931d81e2e03925fe2650231618ebe83f7bab (HEAD -> try/footnotes) wIP
c08cf1a02e14bb95bf37bee734ce35e591c3896e WIP
c55d502f53d6fa561b8eea2424e411b56d8da401 Move footnotes reducer under withChangeDetection
a40b3ee2dbdd18e63421c19d06757be09a932853 Fix sync on DOM update
f986c28fc25488142467621452c6285df229124f Fix e2e tests
c190d175be0095555e8c274f44d1ed77f9215c94 Fix linting
acc7d02fe6bcde076ded94708ee910c98d7deaf3 Allow different footnote blocks
f0456448cf6c36509967190d8f1c8757da2b79a1 Clean up, adjust tests
e697f7d3754cd3c9959d486c5cbac615ebfafabb Clean up
a63e322eac5e0cf9f12f5b2db1ddeaa3a4cb49a9 Clean up selector
e3dbd2fd6dcdee88a359a067d37a74a016c62dce Move styles
0e6c0f031df54b491bcdc495b5d0e74369d367d4 Split edit component
a9278e37b8ad9dfa23eafa1b980b4f5823326d5f Save footnotes in the post content
a473bfb9a4656587fe6929b079b800a06a0e97b7 WIP
9d8b75922f4d5f611e960f438116634723b6357f WIP

ellatrix committed:
Fix sticking on select

ellatrix committed:
WIP: Add selected footnote UI

ellatrix committed:
Hook up delete button

ellatrix commented:
@mcsf I would say that the footnotes list definitely uses the extension mechanisms we have for blocks, but that the list is _not_ a block. To me it sounds ok to display it in the outline, but it will need to be added separately like everywhere else as it's not part of the normal block list.

ellatrix committed:
Remove should account for both types of footnotes

ellatrix commented:
> Does it make sense to be able to add a link to a Footnote? Doesn't it already link to the footnote?

This is a bigger rich text issue. We don't deactivate buttons based on wether clicking them would produce invalid HTML. I would like this as it is right now and look into it separately.

ellatrix commented:
> Can we show the entire footnote in the editor? I was surprised to see it cut off.

I'll now attempt to use `RichText` for these.

ellatrix commented on packages/editor/src/components/footnotes/edit.js:
```diff
@@ -0,0 +1,128 @@
+	withSafeTimeout,
+	withSelect( ( select ) => {
+		const { getEditorBlocks, getFootnotes } = select( 'core/editor' );
+		const { getSelectionStart } = select( 'core/block-editor' );
```
Ideally we use the `core/editor` store here, but we'll need #16428 for that.

ellatrix committed:
Remove selectionchange listener

ellatrix commented:
I'm stumbling upon using `RichText` out of block context, still. For some reason, I cannot let it receive selection. I've been digging around a bit, but I'm not sure I can fix it right now.

Alternatively, we could just a plain `textarea` for now, until we resolve the issues with `RichText outside of a block.

ellatrix commented:
Ok, the problem with `RichText` is that the `Autocomplete` component is still part of it, which causes `Editable` to rerender all the time. This will be fixed by #16905.

ellatrix committed:
Use RichText

ellatrix committed:
Correctly wrap rich text

ellatrix commented:
> Ok, the problem with `RichText` is that the `Autocomplete` component is still part of it, which causes `Editable` to rerender all the time. This will be fixed by #16905.

I changed the base to that PR now. `RichText` has now been added.

ellatrix committed:
Use bare RichText

Soean commented on packages/block-library/src/footnotes/edit.js:
```diff
@@ -0,0 +1,102 @@
+				<>
+					<Toolbar>
+						<IconButton icon="editor-ol" onClick={ viewAll }>
+							View all Footnotes
```
i18n is missing:

```suggestion
							__( 'View all Footnotes' );
```

ellatrix replied:
```suggestion
							{ __( 'View all Footnotes' ) }
```

ellatrix committed:
WIP

ellatrix committed:
Update packages/block-library/src/footnotes/edit.js

Co-Authored-By: Sören Wrede <soerenwrede@gmail.com>

ellatrix commented on packages/block-library/src/footnotes/edit.js:
```diff
@@ -0,0 +1,114 @@
+				<>
+					<Toolbar>
+						<IconButton icon="editor-ol" onClick={ viewAll }>
+							__( 'View all Footnotes' )
```
```suggestion
							{ __( 'View all Footnotes' ) }
```

ellatrix committed:
Update packages/block-library/src/footnotes/edit.js

mcsf commented:
The progress in this PR is very encouraging! Let us know when there's more to review.

Soean commented on packages/format-library/src/footnote/index.js:
```diff
@@ -0,0 +1,104 @@
+			/>
+			<Fill name="__unstable-footnote-controls">
+				<IconButton icon="trash" onClick={ rm }>
+					Remove Footnote
```
i18n is missing.

karmatosed commented:
I tried to run this to give a little design review but it didn't seem to work. Once it is ready for a review please let reach out as excited to see this merged.

ellatrix commented on packages/editor/src/store/reducer.js:
```diff
@@ -142,18 +142,34 @@ export const editor = flow( [
+			resetTypes: [ 'SETUP_EDITOR_STATE', 'REQUEST_POST_UPDATE_START' ],
+		} ),
+	] )( {
+		value( state = [], action ) {
```
@epiqueras Since there is no `editor` recuser anymore, could you point me in the right direction where blocks are (and footnotes could be) stored with change detection and history?

epiqueras replied:
https://github.com/WordPress/gutenberg/blob/79d28866952ec3976eafad41d90458aa9d266455/packages/core-data/src/reducer.js#L154

https://github.com/WordPress/gutenberg/blob/79d28866952ec3976eafad41d90458aa9d266455/packages/core-data/src/reducer.js#L306

LukaszJaro commented:
In the meantime if anyone is looking for a really good solution or is sticking to classic WordPress editor, there is this plugin: https://wordpress.org/plugins/civil-footnotes/ from a WP core contributor, works really well and easy to use.

paaljoachim commented:
@ellatrix can we get a status update on this PR? As the movement seems to have stopped up. 
Thank you.

Soean commented:
@ellatrix We like to help on this PR because we need a footnote block for a project. Can we do anything to push the development or should we start testing?

ellatrix commented:
@Soean I need to rebase this PR first. :) Hoping to do it sometime soon.

mcsf commented:
As a design data point, I like what _Le Monde Diplomatique_ does:

![footnotes-diplo](https://user-images.githubusercontent.com/150562/69821574-fef2a600-11fb-11ea-91a2-2847504e8204.gif)

(there are still actual footnotes at the end of the article, but there are these floating notes too corresponding to visible anchors)

LukaszJaro commented:
Would this feature land in the next major WP version in march?

ellatrix commented:
Unfortunately, I don't think it will make the next WP release. It's a bit too late to work out the rest before beta 1.

mapk commented:
@ellatrix Looks like we've missed the 5.4 release. Is this something we can move on again and get it into the plugin for testing?

Ismail-elkorchi commented:
> @ellatrix Looks like we've missed the 5.4 release. Is this something we can move on again and get it into the plugin for testing?

@mapk This would be indeed a long overdue. Hope it will make it in the 5.5 release.

SpinnerWebs commented:
@ellatrix Is there anything we can do to help with development or testing? Would like to use the feature; am loath to use a different footnote plugin if the facility may shortly be available in Gutenberg.

romainmenke commented:
Any news on this 🙂 
We noticed that https://github.com/WordPress/gutenberg/pull/17824 was merged.
Maybe this is unblocked now?

paaljoachim commented:
@ellatrix It would be nice to get this Foot notes PR added to WP 5.6. I went ahead and added it to the project board: https://github.com/WordPress/gutenberg/projects/48#card-45875045

ellatrix commented:
I'll rework this PR. Have some alternative ideas.

paaljoachim commented:
@ellatrix 
Do you want to close this PR and use information from here in a new PR?

alevals commented:
> ant to close this PR and use information from here in a new PR?

Hello @ellatrix do you have any estimate on when this feature will be released?

ellatrix commented:
Probably WP 5.7

ellatrix commented:
Closing this in favour of #28261.
